### PR TITLE
BUG: Concat multiple different ExtensionArray types

### PR DESCRIPTION
**Pull request by TomAugspurger**
xref https://github.com/pandas-dev/pandas/issues/22994

This builds on https://github.com/pandas-dev/pandas/pull/22996, since we need hashing.

2a1660c77e8ad0c28485c6d81227ac9d6b6de4f2 is the relevant commit.

Sparse tests will fail for now. I'll revisit once SparseArray is in.

### CODE REVIEW

File: pandas/core/dtypes/base.py [hunk 1]
```diff
@@ -40,11 +43,19 @@ def __eq__(self, other):
...
-            return False
+            try:
+                other = self.construct_from_string(other)
+            except TypeError:
+                return False
+        if isinstance(other, type(self)):
+            return all(
+                getattr(self, attr) == getattr(other, attr)
+                for attr in self._metadata
+            )
```
Comment (jbrockmendel) [inline pandas/core/dtypes/base.py]: Corner case: other is an instance of a subclass of type(self) and self._metadata != other._metadata

File: pandas/core/dtypes/concat.py [hunk 2]
```diff
@@ -560,11 +560,6 @@ def _concat_sparse(to_concat, axis=0, typs=None):
 
     fill_values = [x.fill_value for x in to_concat
                    if isinstance(x, SparseArray)]
-
-    if len(set(fill_values)) > 1:
```
Comment (TomAugspurger) [inline pandas/core/dtypes/concat.py]: I think this was leftover from an early implementation that didn't handle multiple fill values. And I think @jorisvandenbossche noticed this in his review, but I misunderstood.

Regardless, we do allow concating Series[sparse] for different fill_values, we just change them all to have the first one's fill value (in `SparseSeries._concat_same_type`)

**Comment by TomAugspurger** [comment]
> Corner case: other is an instance of a subclass of type(self) and self._metadata != other._metadata

This relates to the commits adding hashing (which I was building on top of).

I think that's probably up to the EA author to define a `__eq__` that's correct. We can only do so much.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by TomAugspurger** [comment]
I had to skip some decimal tests on py2.

I couldn't conditionally skip individual tests that had used fixtures defined on the test, e.g. `TestReshaping.test_concat_all_na_block[False]`, since that `in_frame` fixture isn't available to the subclass.

So I opted to skip all the tests on Py2 that had at least one test case failing.

File: doc/source/whatsnew/v0.24.0.txt [hunk 3]
```diff
@@ -602,6 +602,7 @@ update the ``ExtensionDtype._metadata`` tuple to match the signature of your
 - :meth:`Series.astype` and :meth:`DataFrame.astype` now dispatch to :meth:`ExtensionArray.astype` (:issue:`21185:`).
 - Slicing a single row of a ``DataFrame`` with multiple ExtensionArrays of the same type now preserves the dtype, rather than coercing to object (:issue:`22784`)
 - Added :meth:`pandas.api.types.register_extension_dtype` to register an extension type with pandas (:issue:`22664`)
+- Bug in concatenation an Series with two different extension dtypes not casting to object dtype (:issue:`22994`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: double back ticks

File: pandas/tests/extension/decimal/test_decimal.py [hunk 4]
```diff
@@ -174,7 +177,8 @@ class TestCasting(BaseDecimal, base.BaseCastingTests):
 
 
 class TestGroupby(BaseDecimal, base.BaseGroupbyTests):
-    pass
+    pytestmark = pytest.mark.skipif(compat.PY2,
```
Comment (jreback) [inline pandas/tests/extension/decimal/test_decimal.py]: dont' you have a ``@unhashable`` decorator for this?

Comment (TomAugspurger) [reply pandas/tests/extension/decimal/test_decimal.py]: Not sure if we can use it. I've only used that as a regular `skip`, not as a pytestmark or as a `skipif`

**Comment by TomAugspurger** [comment]
I think this is good to go. Merging later today.

FYI, I'm leaving https://github.com/pandas-dev/pandas/issues/22994 open to discuss a more general concat protocol so that types can negotiate on what should happen. I'm hopeful that will clean up all the special concats we have like `concat_sparse` and `concat_categorical`.

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**